File: Kernel Simulator/Misc/Timers/StopwatchScreen.vb

'    Kernel Simulator  Copyright (C) 2018-2022  Aptivi
'
'    This file is part of Kernel Simulator
'
'    Kernel Simulator is free software: you can redistribute it and/or modify
'    it under the terms of the GNU General Public License as published by
'    the Free Software Foundation, either version 3 of the License, or
'    (at your option) any later version.
'
'    Kernel Simulator is distributed in the hope that it will be useful,
'    but WITHOUT ANY WARRANTY; without even the implied warranty of
'    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
'    GNU General Public License for more details.
'
'    You should have received a copy of the GNU General Public License
'    along with this program.  If not, see <https://www.gnu.org/licenses/>.

Imports System.Text
Imports System.Threading

Namespace Misc.Timers
    Public Module StopwatchScreen

        Friend Laps As New List(Of LapDisplayInfo)
        Friend StopwatchUpdate As New KernelThread("Stopwatch ETA Updater", True, AddressOf UpdateStopwatchElapsedDisplay)
        Friend LapColor As Color = NeutralTextColor
        Friend Stopwatch As New Stopwatch
        Friend LappedStopwatch As New Stopwatch
        Friend NewLapAcknowledged As Boolean

        ''' <summary>
        ''' Opens the stopwatch screen
        ''' </summary>
        Sub OpenStopwatch()
            'Clear for cleanliness
            Console.Clear()
            Console.CursorVisible = False

            'Populate the positions for time
            Dim LapsText As String = DoTranslation("Lap")
            Dim HalfWidth As Integer = Console.WindowWidth / 2
            Dim HalfHeight As Integer = Console.WindowHeight / 2
            Dim TimeLeftPosition As Integer = HalfWidth * 1.5 - Stopwatch.Elapsed.ToString("d\.hh\:mm\:ss\.fff", CurrentCult).Length / 2
            Dim TimeTopPosition As Integer = HalfHeight - 2
            Dim LapsCurrentLapLeftPosition As Integer = 4
            Dim LapsCurrentLapTopPosition As Integer = Console.WindowHeight - 6

            'Populate the keys text variable
            Dim KeysText As String = "[ENTER] " + DoTranslation("Start or stop") + " | [L] " + DoTranslation("Lap") + " | [R] " + DoTranslation("Reset") + " | [ESC] " + DoTranslation("Exit")
            Dim KeysTextLeftPosition As Integer = HalfWidth - (KeysText.Length / 2)
            Dim KeysTextTopPosition As Integer = Console.WindowHeight - 2
            Dim KeysKeypress As ConsoleKey

            'Print the keys text
            WriteWhere(KeysText, KeysTextLeftPosition, KeysTextTopPosition, True, ColTypes.Tip)

            'Print the time interval and the current lap
            WriteWhere(Stopwatch.Elapsed.ToString("d\.hh\:mm\:ss\.fff", CurrentCult), TimeLeftPosition, TimeTopPosition, True, LapColor)
            WriteWhere(LapsText + " {0}: {1}", LapsCurrentLapLeftPosition, LapsCurrentLapTopPosition, True, LapColor, Laps.Count + 1, LappedStopwatch.Elapsed.ToString("d\.hh\:mm\:ss\.fff", CurrentCult))

            'Print the border
            MakeBorder()

            While KeysKeypress <> ConsoleKey.Escape
                'Wait for a keypress
                KeysKeypress = DetectKeypress().Key

                'Check for a keypress
                Select Case KeysKeypress
                    Case ConsoleKey.Enter
                        If Not StopwatchUpdate.IsAlive Then
                            StopwatchUpdate.Start()
                        Else
                            StopwatchUpdate.Stop()
                        End If
                        If LappedStopwatch.IsRunning Then LappedStopwatch.Stop() Else LappedStopwatch.Start()
                        If Stopwatch.IsRunning Then Stopwatch.Stop() Else Stopwatch.Start()
                    Case ConsoleKey.L
                        If LappedStopwatch.IsRunning Then
                            Dim Lap As New LapDisplayInfo(LapColor, LappedStopwatch.Elapsed)
                            Laps.Add(Lap)
                            LappedStopwatch.Restart()
                            NewLapAcknowledged = True

                            'Select random color
                            Dim Randomizer As New Random
                            Dim RedValue As Integer = Randomizer.Next(255)
                            Dim GreenValue As Integer = Randomizer.Next(255)
                            Dim BlueValue As Integer = Randomizer.Next(255)
                            LapColor = New Color(RedValue, GreenValue, BlueValue)
                        End If
                    Case ConsoleKey.R
                        If StopwatchUpdate.IsAlive Then
                            StopwatchUpdate.Stop()
                        End If
                        If LappedStopwatch.IsRunning Then LappedStopwatch.Reset()
                        If Stopwatch.IsRunning Then Stopwatch.Reset()

                        'Clear the laps and the laps list
                        Laps.Clear()
                        For Y As Integer = 1 To LapsCurrentLapTopPosition - 1
                            Console.SetCursorPosition(LapsCurrentLapLeftPosition, Y)
                            ClearLineToRight()
                        Next

                        'Reset the indicators
                        LapColor = NeutralTextColor
                        WriteWhere(DoTranslation("Lap") + " {0}: {1}", LapsCurrentLapLeftPosition, LapsCurrentLapTopPosition, False, LapColor, Laps.Count + 1, LappedStopwatch.Elapsed.ToString("d\.hh\:mm\:ss\.fff", CurrentCult))
                        ClearLineToRight()
                        Console.SetCursorPosition(0, TimeTopPosition)
                        ClearLineToRight()
                        WriteWhere(Stopwatch.Elapsed.ToString("d\.hh\:mm\:ss\.fff", CurrentCult), TimeLeftPosition, TimeTopPosition, False, LapColor)
                        MakeBorder()
                    Case ConsoleKey.Escape
                        If LappedStopwatch.IsRunning Then LappedStopwatch.Reset()
                        If Stopwatch.IsRunning Then Stopwatch.Reset()
                        LapColor = NeutralTextColor
                        If StopwatchUpdate.IsAlive Then StopwatchUpdate.Stop()
                End Select
            End While

            'Clear for cleanliness
            Laps.Clear()
            Console.Clear()
            Console.CursorVisible = True
        End Sub

        ''' <summary>
        ''' Updates the elapsed display for stopwatch
        ''' </summary>
        Private Sub UpdateStopwatchElapsedDisplay()
            'Populate the positions for time and for lap list
            Dim LapsText As String = DoTranslation("Lap")
            Dim HalfWidth As Integer = Console.WindowWidth / 2
            Dim HalfHeight As Integer = Console.WindowHeight / 2
            Dim TimeLeftPosition As Integer = HalfWidth * 1.5 - Stopwatch.Elapsed.ToString("d\.hh\:mm\:ss\.fff", CurrentCult).Length / 2
            Dim TimeTopPosition As Integer = HalfHeight - 2
            Dim LapsCurrentLapLeftPosition As Integer = 4
            Dim LapsCurrentLapTopPosition As Integer = Console.WindowHeight - 6
            Dim LapsLapsListLeftPosition As Integer = 4
            Dim LapsLapsListTopPosition As Integer = 3

            While StopwatchUpdate.IsAlive
                Try
                    'Update the elapsed display
                    WriteWhere(Stopwatch.Elapsed.ToString("d\.hh\:mm\:ss\.fff", CurrentCult), TimeLeftPosition, TimeTopPosition, True, LapColor)
                    WriteWhere(LapsText + " {0}: {1}", LapsCurrentLapLeftPosition, LapsCurrentLapTopPosition, True, LapColor, Laps.Count + 1, LappedStopwatch.Elapsed.ToString("d\.hh\:mm\:ss\.fff", CurrentCult))

                    'Update the laps list if new lap is acknowledged
                    If NewLapAcknowledged Then
                        Dim LapsListEndBorder As Integer = Console.WindowHeight - 10
                        Dim LapsListBuilder As New StringBuilder
                        Dim BorderDifference As Integer = Laps.Count - LapsListEndBorder
                        If BorderDifference < 0 Then BorderDifference = 0
                        For LapIndex As Integer = BorderDifference To Laps.Count - 1
                            Dim Lap As LapDisplayInfo = Laps(LapIndex)
                            LapsListBuilder.AppendLine(Lap.LapColor.VTSequenceForeground + DoTranslation("Lap") + $" {LapIndex + 1}: {Lap.LapInterval.ToString("d\.hh\:mm\:ss\.fff", CurrentCult)}")
                        Next
                        WriteWhere(LapsListBuilder.ToString, LapsLapsListLeftPosition, LapsLapsListTopPosition, True, LapColor)
                        NewLapAcknowledged = False
                    End If
                Catch ex As ThreadInterruptedException
                    Exit While
                End Try
            End While
        End Sub

        ''' <summary>
        ''' Makes the display border
        ''' </summary>
        Sub MakeBorder()
            Dim KeysTextTopPosition As Integer = Console.WindowHeight - 2
            Dim HalfWidth As Integer = Console.WindowWidth / 2
            WriteWhere("═".Repeat(Console.WindowWidth), 0, KeysTextTopPosition - 2, True, ColTypes.Gray)
            WriteWhere("═".Repeat(Console.WindowWidth), 0, 1, True, ColTypes.Gray)
            For Height = 2 To KeysTextTopPosition - 2
                WriteWhere("║", HalfWidth, Height, True, ColTypes.Gray)
            Next
            WriteWhere("╩", HalfWidth, KeysTextTopPosition - 2, True, ColTypes.Gray)
            WriteWhere("╦", HalfWidth, 1, True, ColTypes.Gray)
        End Sub

    End Module
End Namespace

File: Kernel Simulator/Misc/Timers/StopwatchScreen.cs
//
// Kernel Simulator  Copyright (C) 2018-2024  Aptivi
//
// This file is part of Kernel Simulator
//
// Kernel Simulator is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// Kernel Simulator is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY, without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

using System;
using System.Collections.Generic;
using System.Diagnostics;

// Kernel Simulator  Copyright (C) 2018-2022  Aptivi
// 
// This file is part of Kernel Simulator
// 
// Kernel Simulator is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// Kernel Simulator is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

using System.Text;
using System.Threading;
using KS.ConsoleBase.Colors;
using KS.ConsoleBase.Inputs;
using KS.Languages;
using KS.Misc.Text;
using KS.Misc.Threading;
using KS.Misc.Writers.ConsoleWriters;
using Terminaux.Base;
using Terminaux.Colors;

namespace KS.Misc.Timers
{
    public static class StopwatchScreen
    {

        internal static List<LapDisplayInfo> Laps = [];
        internal static KernelThread StopwatchUpdate = new("Stopwatch ETA Updater", true, UpdateStopwatchElapsedDisplay);
        internal static Color LapColor = KernelColorTools.NeutralTextColor;
        internal static Stopwatch Stopwatch = new();
        internal static Stopwatch LappedStopwatch = new();
        internal static bool NewLapAcknowledged;
        internal static bool stopwatchStopping;

        /// <summary>
        /// Opens the stopwatch screen
        /// </summary>
        public static void OpenStopwatch()
        {
            // Clear for cleanliness
            ConsoleWrapper.Clear();
            ConsoleWrapper.CursorVisible = false;

            // Populate the positions for time
            string LapsText = Translate.DoTranslation("Lap");
            int HalfWidth = (int)Math.Round(ConsoleWrapper.WindowWidth / 2d);
            int HalfHeight = (int)Math.Round(ConsoleWrapper.WindowHeight / 2d);
            int TimeLeftPosition = (int)Math.Round(HalfWidth * 1.5d - Stopwatch.Elapsed.ToString(@"d\.hh\:mm\:ss\.fff", CultureManager.CurrentCult).Length / 2d);
            int TimeTopPosition = HalfHeight - 2;
            int LapsCurrentLapLeftPosition = 4;
            int LapsCurrentLapTopPosition = ConsoleWrapper.WindowHeight - 6;

            // Populate the keys text variable
            string KeysText = "[ENTER] " + Translate.DoTranslation("Start or stop") + " | [L] " + Translate.DoTranslation("Lap") + " | [R] " + Translate.DoTranslation("Reset") + " | [ESC] " + Translate.DoTranslation("Exit");
            int KeysTextLeftPosition = (int)Math.Round(HalfWidth - KeysText.Length / 2d);
            int KeysTextTopPosition = ConsoleWrapper.WindowHeight - 2;
            var KeysKeypress = default(ConsoleKey);

            // Print the keys text
            TextWriterWhereColor.WriteWhere(KeysText, KeysTextLeftPosition, KeysTextTopPosition, true, KernelColorTools.ColTypes.Tip);

            // Print the time interval and the current lap
            TextWriterWhereColor.WriteWhere(Stopwatch.Elapsed.ToString(@"d\.hh\:mm\:ss\.fff", CultureManager.CurrentCult), TimeLeftPosition, TimeTopPosition, true, LapColor);
            TextWriterWhereColor.WriteWhere(LapsText + " {0}: {1}", LapsCurrentLapLeftPosition, LapsCurrentLapTopPosition, true, LapColor, vars: [Laps.Count + 1, LappedStopwatch.Elapsed.ToString(@"d\.hh\:mm\:ss\.fff", CultureManager.CurrentCult)]);

            // Print the border
            MakeBorder();

            while (KeysKeypress != ConsoleKey.Escape)
            {
                // Wait for a keypress
                KeysKeypress = Input.DetectKeypress().Key;

                // Check for a keypress
                switch (KeysKeypress)
                {
                    case ConsoleKey.Enter:
                        {
                            if (!StopwatchUpdate.IsAlive)
                            {
                                stopwatchStopping = false;
                                StopwatchUpdate.Start();
                            }
                            else
                            {
                                stopwatchStopping = true;
                                StopwatchUpdate.Stop();
                            }
                            if (LappedStopwatch.IsRunning)
                                LappedStopwatch.Stop();
                            else
                                LappedStopwatch.Start();
                            if (Stopwatch.IsRunning)
                                Stopwatch.Stop();
                            else
                                Stopwatch.Start();
                            break;
                        }
                    case ConsoleKey.L:
                        {
                            if (LappedStopwatch.IsRunning)
                            {
                                var Lap = new LapDisplayInfo(LapColor, LappedStopwatch.Elapsed);
                                Laps.Add(Lap);
                                LappedStopwatch.Restart();
                                NewLapAcknowledged = true;

                                // Select random color
                                var Randomizer = new Random();
                                int RedValue = Randomizer.Next(255);
                                int GreenValue = Randomizer.Next(255);
                                int BlueValue = Randomizer.Next(255);
                                LapColor = new Color(RedValue, GreenValue, BlueValue);
                            }

                            break;
                        }
                    case ConsoleKey.R:
                        {
                            if (StopwatchUpdate.IsAlive)
                            {
                                stopwatchStopping = true;
                                StopwatchUpdate.Stop();
                            }
                            if (LappedStopwatch.IsRunning)
                                LappedStopwatch.Reset();
                            if (Stopwatch.IsRunning)
                                Stopwatch.Reset();

                            // Clear the laps and the laps list
                            Laps.Clear();
                            for (int Y = 1, loopTo = LapsCurrentLapTopPosition - 1; Y <= loopTo; Y++)
                            {
                                ConsoleWrapper.SetCursorPosition(LapsCurrentLapLeftPosition, Y);
                                ConsoleBase.ConsoleExtensions.ClearLineToRight();
                            }

                            // Reset the indicators
                            LapColor = KernelColorTools.NeutralTextColor;
                            TextWriterWhereColor.WriteWhere(Translate.DoTranslation("Lap") + " {0}: {1}", LapsCurrentLapLeftPosition, LapsCurrentLapTopPosition, false, LapColor, vars: [Laps.Count + 1, LappedStopwatch.Elapsed.ToString(@"d\.hh\:mm\:ss\.fff", CultureManager.CurrentCult)]);
                            ConsoleBase.ConsoleExtensions.ClearLineToRight();
                            ConsoleWrapper.SetCursorPosition(0, TimeTopPosition);
                            ConsoleBase.ConsoleExtensions.ClearLineToRight();
                            TextWriterWhereColor.WriteWhere(Stopwatch.Elapsed.ToString(@"d\.hh\:mm\:ss\.fff", CultureManager.CurrentCult), TimeLeftPosition, TimeTopPosition, false, LapColor);
                            MakeBorder();
                            break;
                        }
                    case ConsoleKey.Escape:
                        {
                            if (LappedStopwatch.IsRunning)
                                LappedStopwatch.Reset();
                            if (Stopwatch.IsRunning)
                                Stopwatch.Reset();
                            LapColor = KernelColorTools.NeutralTextColor;
                            if (StopwatchUpdate.IsAlive)
                            {
                                stopwatchStopping = true;
                                StopwatchUpdate.Stop();
                            }
                            break;
                        }
                }
            }

            // Clear for cleanliness
            Laps.Clear();
            ConsoleWrapper.Clear();
            ConsoleWrapper.CursorVisible = true;
            stopwatchStopping = false;
        }

        /// <summary>
        /// Updates the elapsed display for stopwatch
        /// </summary>
        private static void UpdateStopwatchElapsedDisplay()
        {
            // Populate the positions for time and for lap list
            string LapsText = Translate.DoTranslation("Lap");
            int HalfWidth = (int)Math.Round(ConsoleWrapper.WindowWidth / 2d);
            int HalfHeight = (int)Math.Round(ConsoleWrapper.WindowHeight / 2d);
            int TimeLeftPosition = (int)Math.Round(HalfWidth * 1.5d - Stopwatch.Elapsed.ToString(@"d\.hh\:mm\:ss\.fff", CultureManager.CurrentCult).Length / 2d);
            int TimeTopPosition = HalfHeight - 2;
            int LapsCurrentLapLeftPosition = 4;
            int LapsCurrentLapTopPosition = ConsoleWrapper.WindowHeight - 6;
            int LapsLapsListLeftPosition = 4;
            int LapsLapsListTopPosition = 3;

            while (!stopwatchStopping)
            {
                try
                {
                    // Update the elapsed display
                    TextWriterWhereColor.WriteWhere(Stopwatch.Elapsed.ToString(@"d\.hh\:mm\:ss\.fff", CultureManager.CurrentCult), TimeLeftPosition, TimeTopPosition, true, LapColor);
                    TextWriterWhereColor.WriteWhere(LapsText + " {0}: {1}", LapsCurrentLapLeftPosition, LapsCurrentLapTopPosition, true, LapColor, vars: [Laps.Count + 1, LappedStopwatch.Elapsed.ToString(@"d\.hh\:mm\:ss\.fff", CultureManager.CurrentCult)]);

                    // Update the laps list if new lap is acknowledged
                    if (NewLapAcknowledged)
                    {
                        int LapsListEndBorder = ConsoleWrapper.WindowHeight - 10;
                        var LapsListBuilder = new StringBuilder();
                        int BorderDifference = Laps.Count - LapsListEndBorder;
                        if (BorderDifference < 0)
                            BorderDifference = 0;
                        for (int LapIndex = BorderDifference, loopTo = Laps.Count - 1; LapIndex <= loopTo; LapIndex++)
                        {
                            var Lap = Laps[LapIndex];
                            LapsListBuilder.AppendLine(Lap.LapColor.VTSequenceForeground + Translate.DoTranslation("Lap") + $" {LapIndex + 1}: {Lap.LapInterval.ToString(@"d\.hh\:mm\:ss\.fff", CultureManager.CurrentCult)}");
                        }
                        TextWriterWhereColor.WriteWhere(LapsListBuilder.ToString(), LapsLapsListLeftPosition, LapsLapsListTopPosition, true, LapColor);
                        NewLapAcknowledged = false;
                    }
                }
                catch (ThreadInterruptedException)
                {
                    break;
                }
            }
        }

        /// <summary>
        /// Makes the display border
        /// </summary>
        public static void MakeBorder()
        {
            int KeysTextTopPosition = ConsoleWrapper.WindowHeight - 2;
            int HalfWidth = (int)Math.Round(ConsoleWrapper.WindowWidth / 2d);
            TextWriterWhereColor.WriteWhere("═".Repeat(ConsoleWrapper.WindowWidth), 0, KeysTextTopPosition - 2, true, KernelColorTools.ColTypes.Gray);
            TextWriterWhereColor.WriteWhere("═".Repeat(ConsoleWrapper.WindowWidth), 0, 1, true, KernelColorTools.ColTypes.Gray);
            for (int Height = 2, loopTo = KeysTextTopPosition - 2; Height <= loopTo; Height++)
                TextWriterWhereColor.WriteWhere("║", HalfWidth, Height, true, KernelColorTools.ColTypes.Gray);
            TextWriterWhereColor.WriteWhere("╩", HalfWidth, KeysTextTopPosition - 2, true, KernelColorTools.ColTypes.Gray);
            TextWriterWhereColor.WriteWhere("╦", HalfWidth, 1, true, KernelColorTools.ColTypes.Gray);
        }

    }
}
